man/man1/regridsoil.1
.TH regridsoil 1  "USER COMMANDS"
.SH NAME
regridsoil \- regrid soil file to new grid
.SH SYNOPSIS
.B regridsoil
[\-h] [\-search] [\-zero] [\-longheader] [\-json]
.I grid_old.clm grid_new.clm soil_old.clm soil_new.clm
.SH DESCRIPTION
Program reads soil file and corresponding grid file to regrid the soil data to a new grid file.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-search
If grid coordinate cannot be found in old grid nearest grid cell is used.
.TP
\-zero
If grid coordinate cannot be found in old grid soil code is set to zero.
.TP
\-longheader
Version of CLM grid file is set to 2. By default version is read from file header.
.TP
\-json
An additional JSON metafile with suffix \fI.json\fP is written.
.TP
.I grid_old.clm
Corresponding grid filename of input data.
.TP
.I grid_new.clm
New grid filename data should be regridded to.
.TP
.I  soil_old.clm
Filename of soil file that should be regridded.
.TP
.I soil_new.clm
Filename of soil file where regridded data is written.
.SH EXAMPLES
.TP
Regrid soil file to the grid file of the amazon basin:
.B regridsoil
grid.clm grid_amazon.clm soil.bin soil_amazon.bin
.SH EXIT STATUS
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
regridlpj(1), regridclm(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review:

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes: 
  - `alphaa` from 0.5 to 0.6 for natural vegetation; 
  - `theta` from 0.9 to 0.95; 
  - `fn_turnover` PFT1: 0.7, PFT2: 0.3, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3;
- parameter `bc3` and `bc4` now computed dependent on mean vegetation period temperature
- nitrogen from turnover directly added to `bm_inc.nitrogen`
- N uptake now separated for uptake of NH4 and NO3 with literature-based parameters

### Fixed

- corrected conversion of vmax in computation of N demand, which was incorrectly scaled with `daylength`
- corrected temperature sensitive of N demand
- corrected `f_NCplant` to comply with equation from [Smith et al. 2014](https://doi.org/10.5194/bg-11-2027-2014)
- corrected phenology of trees...


## [5.9.27] - 2025-07-15

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de)

### Added

- Utility `restart2yaml` added to convert the new restart files into human readable YAML or JSON files.
- Utility `json2restart` added to convert JSON files into restart files.
- alias `restart2json` added for `restart2yaml -json`.
- Output of time spent in reading and writing the restart file added.
- Option `-print_noread` added to `lpjml` to print variable names not read from restart file.
- Check added that PFT names in restart file match PFT names in LPJmL configuration file.
- Unit tests added for bstruct library functions.

### Changed

- Restart file format changed to include metadata. A binary JSON-like structure has been implemented. The order of the objects read can be different from the order objects have been written, but performance of reading can be degraded.
- Utility `lpjcat` updated to support the new restart file format.
- Preprint's doi:10.5194/egusphere-2023-2946 changed to the final BNF paper doi:10.5194/gmd-17-7889-2024 in `.zenodo.json`.


## [5.9.26] - 2025-07-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hötten (davidho@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Changed

- New functions `setupannual_grid`, `initmonthly_grid()`, `updatedaily_grid()` and `updateannual_grid()` defined to minimize lines of code in `iterateyear.c` and `iterate.c` and minimize duplicate code in `lpj_poem.c`.
- Parts of code in `iterateyear.c` has been moved to `update_daily_cell.c` and `update_monthly_grid.c`.
- `lpj_climber4.c` renamed to `lpj_poem.c`.
- `update_daily.c` renamed to `update_daily_cell.c`.
- `update_monthly.c` renamed to `update_monthly_grid.c`.
- `AFTER_STRESS` debugging output in `update_daily_cell.c' (former `updatedaily.c`) moved after N update really happened.

### Fixed

- Macro `isequilyear()` has been defined for calling equilibration function in a consistent way (issue #375).


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.

src/utils/regridsoil.c
/**************************************************************************************/
/**                                                                                \n**/
/**                    r  e  g  r  i  d  s  o  i  l  .  c                          \n**/
/**                                                                                \n**/
/**     soil data is regridded to new grid file                                    \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-longheader] [-search] [-zero] [-json] coord_old.clm coord_new.clm\n       soil_old.bin soil_new.bin\n"
#define ERR_USAGE USAGE "\nTry \"%s --help\" for more information.\n"

int main(int argc,char **argv)
{
  FILE *file;
  Coord *c,*c2;
  Bool iszero,issearch,isjson;
  Byte *soil,zero=0;
  char *arglist,*out_json;
  const char *progname;
  Coordfile grid;
  float lon,lat;
  Coord res,res2;
  Real dist_min;
  int i,j,setversion,ngrid,ngrid2,count,iarg,rc;
  Filename filename,grid_name;
  Type grid_type;
  Header header;
  setversion=READ_VERSION;
  iszero=issearch=isjson=FALSE;
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("     ");
        rc=printf("%s (" __DATE__ ") Help",progname);
        printf("\n     ");
        repeatch('=',rc);
        printf("\n\nRegrid raw binary soil file to a new grid for LPJmL version %s\n",getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h,--help    print this help text\n"
               "-search      if cell is not found, nearest cell is used\n"
               "-zero        id cell is not found, data is set to zero\n"
               "-json        an additional JSON metafile with suffix .json is created\n"
               "grid_old.clm corresponding grid filename of input data\n"
               "grid_new_clm grid filename data should be regridded to\n"
               "soil_old_clm filename of data that should be regridded\n"
               "soil_new.clm filename of data file where regridded data is written\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               progname);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[iarg],"-longheader"))
        setversion=2;
      else if(!strcmp(argv[iarg],"-search"))
        issearch=TRUE;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-zero"))
        iszero=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                ERR_USAGE,argv[iarg],progname,progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;

  if(argc<4+iarg)
  {
    fprintf(stderr,"Missing arguments.\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  filename.name=argv[iarg];
  filename.fmt=(setversion==2) ? CLM2 : CLM;
  grid=opencoord(&filename,TRUE);
  if(grid==NULL)
    return EXIT_FAILURE;
  ngrid=numcoord(grid);
  getcellsizecoord(&lon,&lat,grid);
  res.lon=lon;
  res.lat=lat;
  c=newvec(Coord,ngrid);
  if(c==NULL)
  {
    printallocerr("c");
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid;i++)
    if(readcoord(grid,c+i,&res))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[iarg]);
      return EXIT_FAILURE;
    }
  closecoord(grid);
  filename.name=argv[iarg+1];
  filename.fmt=(setversion==2) ? CLM2 : CLM;
  grid=opencoord(&filename,TRUE);
  if(grid==NULL)
    return EXIT_FAILURE;
  ngrid2=numcoord(grid);
  grid_type=getcoordtype(grid);
  getcellsizecoord(&lon,&lat,grid);
  res2.lon=lon;
  res2.lat=lat;
  if(res.lon!=res2.lon)
    fprintf(stderr,"Warning: longitudinal resolution %g in '%s' differs from %g in '%s.\n",res2.lon,argv[iarg+1],res.lon,argv[iarg]);
  if(res.lat!=res2.lat)
    fprintf(stderr,"Warning: latitudinal resolution %g in '%s' differs from %g in '%s.\n",res2.lat,argv[iarg+1],res.lat,argv[iarg]);
  c2=newvec(Coord,ngrid2);
  if(c2==NULL)
  {
    printallocerr("c2");
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid2;i++)
    if(readcoord(grid,c2+i,&res2))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[iarg+1]);
      return EXIT_FAILURE;
    }
  closecoord(grid);
  file=fopen(argv[iarg+2],"rb");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s.\n",argv[iarg+2],strerror(errno));
    return EXIT_FAILURE;
  }
  soil=malloc(ngrid);
  if(soil==NULL)
  {
    printallocerr("soil");
    return EXIT_FAILURE;
  }
  if(fread(soil,1,ngrid,file)!=ngrid)
  {
    fprintf(stderr,"Error reading '%s': %s.\n",argv[iarg+2],strerror(errno));
    return EXIT_FAILURE;
  }
  fclose(file);
  file=fopen(argv[iarg+3],"wb");
  if(file==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[iarg+3],strerror(errno));
    return EXIT_FAILURE;
  }
  count=0;
  printf("[     ");
  for(i=0;i<ngrid2;i++)
  {
    if(ngrid2>10 && (i %(ngrid2/10)==0))
    {
      printf("\b\b\b\b\b%3d%%]",(i/(ngrid2/10))*10);
      fflush(stdout);
    }
#ifdef DEBUG
    printf("%d ",i);
    printcoord(c2+i);
    fputs(":",stdout);
#endif
    j=findcoord(c2+i,c,&res,ngrid);
    if(j==NOT_FOUND)
    {
      count++;
      if(issearch)
      {
        fputs("Coordinate ",stderr);
        fprintcoord(stderr,c2+i);
        fputs(" not found, replaced by ",stderr);
        j=findnextcoord(&dist_min,c2+i,c,ngrid);
        fprintcoord(stderr,c+j);
        fprintf(stderr,", distance=%g\n",dist_min);
      }
      else
      {
        fputs("Coordinate ",stderr);
        fprintcoord(stderr,c2+i);
        fprintf(stderr," at index %d not found.\n",i);
        j=ngrid;
        if(!iszero)
          return EXIT_FAILURE;
      }
    }
    if(j==ngrid)
      fwrite(&zero,1,1,file);
    else
      fwrite(soil+j,1,1,file);
  }
  putchar('\n');
  free(soil);
  free(c);
  free(c2);
  if(count)
    fprintf(stderr,"Warning: %d cells not found.\n",count);
  fclose(file);
  if(isjson)
  {
    out_json=malloc(strlen(argv[iarg+3])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[iarg+3]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    file=fopen(out_json,"w");
    if(file==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    grid_name.name=argv[iarg+1];
    grid_name.fmt=CLM;
    header.datatype=LPJ_BYTE;
    header.firstcell=0;
    header.ncell=ngrid2;
    header.timestep=header.nstep=header.nbands=1;
    header.firstyear=1901;
    header.nyear=1;
    header.scalar=1;
    header.order=CELLYEAR;
    header.cellsize_lon=lon;
    header.cellsize_lat=lat;
    fprintjson(file,argv[iarg+3],NULL,NULL,NULL,arglist,&header,NULL,NULL,NULL,0,NULL,NULL,NULL,NULL,&grid_name,grid_type,RAW,NULL,FALSE,0);
    free(out_json);
    free(arglist);
    fclose(file);
  }
  return EXIT_SUCCESS;
}  /* of 'main' */
